pkg/packagekit/kpatch.jsx

```diff
@@ -0,0 +1,376 @@
+
+const _ = cockpit.gettext;
+
+export class KpatchSettings extends React.Component {
+    constructor() {
+        super();
+
+        this.state = {
+            auto: null,
+            loaded: false,
+            enabled: null,
+            missing: [],
+            unavailable: [],
+            error: "",
+            updating: false,
+            showModal: false,
+            applyCheckbox: false,
+            justCurrent: null,
+            kernelName: "",
+            patchName: null,
+            patchInstalled: null,
+            patchUnavailable: null,
+        };
+
+        this.kpatchService = serviceProxy("kpatch");
+
+        this.checkSetup = this.checkSetup.bind(this);
+        this.handleChange = this.handleChange.bind(this);
+        this.onClose = this.onClose.bind(this);
+        this.handleInstall = this.handleInstall.bind(this);
+    }
+
+    componentDidMount() {
+        check_missing_packages(["kpatch", "kpatch-dnf"])
+                .then(d => {
+                    this.setState({
+                        loaded: true,
+                        unavailable: d.unavailable_names || [],
+                        missing: d.missing_names || [],
+                    });
+
+                    this.checkSetup();
+                })
+                .catch(e => {
+                    console.log("Could not determine kpatch availability.");
+                });
+
+        this.kpatchService.addEventListener('changed', () => {
+            this.setState((state, _) => {
+                const current = this.kpatchService.enabled && (state.patchInstalled || state.patchUnavailable);
+                return ({
+                    enabled: this.kpatchService.enabled,
+                    justCurrent: current && !state.auto,
+                    applyCheckbox: current,
+                });
+            });
+        });
+    }
+
+    checkSetup() {
+        cockpit.file("/etc/dnf/plugins/kpatch.conf").read()
+                .then(data => {
+                    if (data) {
+                        const auto = /autoupdate\s*=\s*True/.test(data);
+                        this.setState((state, _) => {
+                            const current = state.enabled && (state.patchInstalled || state.patchUnavailable);
+                            return ({
+                                auto: !!auto,
+                                justCurrent: current && !auto,
+                                applyCheckbox: current,
+                            });
+                        });
+                    }
+                })
+                .catch(() => true); // Ignore errors, most likely just does not exist
+
+        cockpit.spawn(["uname", "-r"])
+                .then(data => {
+                    const fields = data.split(".");
+                    const sanitized_kernel_ver = fields.slice(0, fields.length - 2)
+                            .join("_");
+                    const patch_name = "kpatch-patch-" + sanitized_kernel_ver;
```

Comment:
it's not that the full version is "insane", it just includes the distro revision. I suggest to first split it by `-` (upstream vs. downstream), and then handle the upstream version. I am rather surprised that kpatch-patch-* package names don't include the distro revision, though -- these often have hundreds of patches which are relevant to the binary diffs.

Is that naming pattern something which we can rely on? That feels like a pretty big assumption. This is something which I complained about in the earliest design phase already -- there should be some metapackage which pulls in the kpatch-patches-version ones, or the `dnf kpatch` command should do that. We don't want admins having to fish around for such package version numbers. Cockpit may be able to figure these out (until they change the naming pattern), but humans?

Reply:
https://github.com/dynup/kpatch-dnf/blob/master/kpatch.py#L50 - we can use this same pattern then.

> or the dnf kpatch command should do that

It does it, but only when you do `dnf kpatch auto`. But that also enables future updates. This was the biggest change with the new design. Maybe if we could teach `dnf kpatch` `install` command that would only install patch but not subscribe to patches for future kernels it would simplify things a lot. Also that should be a simple patch for kpatch-dnf.

Reply:
https://github.com/dynup/kpatch-dnf/pull/8

Reply:
Actually, if we are going to delegate more things to kpatch-dnf, lets also move there logic for recognizing if there is something to be installed in the first place.